src/library/scala/collection/mutable/ArrayBuilder.scala

```diff
@@ -520,3 +531,6 @@ object ArrayBuilder {
     override def toString = "ArrayBuilder.ofUnit"
   }
 }
+
+class ExceededMaximumArrayBuilderCapacity extends Exception(s"ArrayBuilder cannot store more than ${PStatics.VM_MaxArraySize} elements.")
+class EncounteredNegativeArrayBuilderCapacity(size:Int) extends Exception(s"ArrayBuilder encountered negative size parameter: $size.  This may have resulted from an overflow.")
```

Comment:
Why not just try to allocate the array and let it fail there with a `NegativeArraySizeException`?

Reply:
If you look at ensureSize, you'll see that negative values can trigger infinite loops.  Programs just hang without any errors or exceptions when trying to expand an array beyond 1073741824.

My next submission will defer to natural exceptions and focus more on preventing the infinite loop.